### ISSUE TITLE
QR Code Reading seems to have an unspoken limit - either based on total size of all code contents or number of codes

### ISSUE THREAD

**Issue by BenWoodford** [microsoft/MixedRealityToolkit-Unity#3146]
I have a few things I need to test on this still but wanted to write down what I’ve found. I was playing around with the QR reading and had an issue reading one of the codes, it would fail at the point where it gets the `relativePose` to position the code object, for some unknown reason that value would just come back null even though the code was right in front of me.

I restarted MRP and replugged the HMD, tried again reading codes in a different order and at some point a few codes (I had added more codes to the room at this point) had the same issue. They had an ID, had contents, but there were no spatial coordinates for them and `relativePose` was null.

I have a few codes that are full-on blocks of JSON, as I was testing how high a version I could get to read (answer is only the first few sizes of QR codes will read) so I’m wondering if whatever is storing the spatial coordinates is filling up when storing these extra large codes. The other option is there is a limit on the number of codes you can track before it stops reading new ones, but if so this limit isn’t terribly big.

I’m aiming to get some exact numbers so I can give some more precise answers. If anyone wants to print off a whole bunch of codes you can do so here: https://qr.fenneclabs.co.uk - the print view will segment them 1 to a page and even gives you the code that they contain.

**Comment by BenWoodford** [comment]
CC @chgatla-microsoft and @yoyozilla

**Comment by chgatla-microsoft** [comment]
Current version of QR codes reads only default encoding for bytes and numeric and alphanumeric ones. The future API will also decode other encodings. If for some reason the API is not able to decode the "code" will be null string, in this case the user can parse the stream to make sense of the code. ( I would not recommend doing this as platform will update the plugin to handle more encodings soon, rather use codes with default encoding and do no not use UTF8 or Kanji.)
Spatial Coordinate system will not be available if the QR code is not currently locatable. For example tracking was lost and it should be available as soon as they are relocated in the space. There is a limit of 100 QR codes, when it hits this it will start filtering out older ones. This limit has nothing to do with the spatial coordinate system. We allow up to 20 version, i.e anything > 20 will surely not be detected. And practically the detectable versions are < 7 depending on the size you print. Anything in between 7 - 20 is unpredictable. If you have default encoding QR codes that are not detected please attach the samples so we can take a look at.

**Comment by BenWoodford** [comment]
@chgatla-microsoft you misunderstand, the issue I’m coming across is that after a handful of codes (6 or so?) it will start failing to get codes. These are codes that are actually being detected, but the relativePose returns null when fetched. The code in question is not consistent, it reads the code fine, but fails to get the updated position (it’s null)

I’m going to try and get reproduction steps, these are all strings so no weird encoding or anything like that, and when I restart the driver the code that fails will change. It’s as if it’s running out but if it’s 100 codes that would be weird.

Shall try and get steps for you tomorrow.

**Comment by BenWoodford** [comment]
So I think what I've actually found here is just an issue with the cameras. They're not terribly good in bright lights. We have quite bright fluorescent lighting in the office and when you're looking straight down at very white paper on a white desk the headset can barely make out the code at times. Still, it seems odd that it fires an update for that QR code when it doesn't have coordinates for it - is there a possibility of it being able to read a code but not able to position it in 3D space and that's when this instance pops up?

Also, any way to improve the camera quality at all? It looks pretty gnarly through flashlight which is how I worked out that it wasn't able to focus on the QR codes correctly.

I had suggested ArUco codes to @Yoyozilla in an email, may be worth looking into as they're a lot less complex and have bigger squares when you're dealing with low numbers, so it could be a good tracking target alongside QR codes if you just need to store very small amounts of data, and it would be a hell of a lot easier for the cameras to capture.
